RPG-playerstats: global stats and add/display functions

sumedhsarwate opened the pull request:
We've implemented the global stats struct and the add_stat_player and display_stat functions and tests. Addresses issues #591 and #609.

carlosazpurua1004 commented:
If you look on the travis output
https://travis-ci.com/github/uchicago-cs/chiventure/builds/168145141
You can see that they didn't actually run the rpg-stats tests (I think).

sumedhsarwate commented:
@carlosazpurua1004 the rpg-stat tests are a part of the game-state tests (the tests for add and display are 2 out of the 101 tests that were run)

ecoble commented on include/game-state/stats.h:
```diff
@@ -3,10 +3,27 @@
+  * */
+typedef struct stats_global{
+    char *name;
+    double max;
```
If the global stats has a max, should player stat also have a max? What is the difference between these two?

sumedhsarwate replied:
I believe the global stats max is the max value a stat could ever have, while the player stat max is the current max for the player/npc and it could be dependent on things like class or level, but it'll never be greater than the global max.

ecoble commented on include/game-state/stats.h:
```diff
@@ -169,14 +215,25 @@ int add_stat_player(stats_hash_t *sh, stats_t *s);
  *
  * Parameters: 
- * sh: pointer to the stats hash table to be freed
+ * gsh: pointer to the global statto be freed
```
The parameter is called stat, also need a space between "stat" and "to"

ecoble commented on src/game-state/src/stats.c:
```diff
@@ -43,17 +43,42 @@ double get_stat_mod(stats_hash_t *sh, char *stat)
-    printf("display_stats: function not yet implemented\n");
-    return "0"; // still needs to be implemented
+    stats_t *stat;
+    int size = 2 + (20 * HASH_COUNT(s));
```
Create macros for 2 and 20, no magic numbers

ecoble commented on include/game-state/stats.h:
```diff
@@ -66,30 +83,59 @@ typedef struct effects{
+ *  Pointer to allocated global stats struct
+ */
+
+stats_global_t* stats_global_new(char *name, double max);
```
move *

ecoble commented on tests/game-state/test_stats.c:
```diff
@@ -0,0 +1,70 @@
+Test (stats, add_stat_player)
+{
+    stats_hash_t *sh = NULL;
+    stats_global_t g;
```
Once stats_global_new gets implemented, you should change this test to create a global stat using that function

sumedhsarwate commented:
@ecoble @carlosazpurua1004 I just realized that the header comment for display_stats says the function should return the names of the stats along with their values/modifiers, but I currently have it as only returning a list of their names. Should I change the function so it returns a string in a format like "stat name: value, modifier\n" or leave it as is?

ecoble commented:
> @ecoble @carlosazpurua1004 I just realized that the header comment for display_stats says the function should return the names of the stats along with their values/modifiers, but I currently have it as only returning a list of their names. Should I change the function so it returns a string in a format like "stat name: value, modifier\n" or leave it as is?

It's up to you, I think leaving it is fine since I'm not sure how much this function will end up being used anyways

ecoble commented on include/game-state/stats.h:
```diff
@@ -3,21 +3,38 @@
+  *
+  *      the global maximal value a stat could have
+  * */
+typedef struct stats_global{
```
Space before brace

ecoble commented on include/game-state/stats.h:
```diff
@@ -33,7 +50,7 @@ typedef struct stats stats_hash_t;
  *      the modifier of the effect on that stat 
  * 
- *      the duration of the effect, an umber 
+ *      the duration of the effect, an number 
```
nit: a number

ecoble replied:
Although really it would be better to say "an int"

ecoble commented on include/game-state/stats.h:
```diff
@@ -66,30 +83,59 @@ typedef struct effects{
  *
  * Parameters:
- * stats_name: the unique string ID to be given to the stat
+ * stats_name: the pointer to the global stat struct.
```
The parameter is named "stat"

ecoble commented on include/game-state/stats.h:
```diff
@@ -155,7 +201,7 @@ double get_stat_mod(stats_hash_t *sh, char *stat);
+int add_stat_player(stats_hash_t **sh, stats_t *s);
 
 /*
  * Print the stats in a hashtable and their values/modifiers
```
As you commented, this can be changed to just print the stat names

ecoble commented on include/game-state/stats.h:
```diff
@@ -169,14 +215,25 @@ int add_stat_player(stats_hash_t *sh, stats_t *s);
+ * Frees a stat
+ *
+ * Parameters: 
+ * sh: pointer to the stat to be freed
```
The parameter is named stat

ecoble commented on include/game-state/stats.h:
```diff
@@ -169,14 +215,25 @@ int add_stat_player(stats_hash_t *sh, stats_t *s);
  *  SUCCESS on success, FAILURE if an error occurs.
  */
-int free_stats(stats_hash_t *sh);
+int free_stats_global(stats_global_t* stat);
```
move * to be stats_global_t *stat

ecoble commented on src/game-state/src/stats.c:
```diff
@@ -1,14 +1,16 @@
 #include "game-state/stats.h"
+#define MIN_STRING_LENGTH 2
+#define MAX_NAME_LENGTH 20
```
Consider upping this to 40 or 50, just to have a buffer

sumedhsarwate commented:
I'm going to keep the implementation for display_stat as it is, so I've updated the header comment for it and I've addressed the other comments as well.

carlosazpurua1004 commented on include/game-state/stats.h:
```diff
@@ -169,14 +215,25 @@ int add_stat_player(stats_hash_t *sh, stats_t *s);
 char *display_stats(stats_hash_t *sh);
 
 /*
- * Frees a stats hash table
+ * Frees a stat
+ *
+ * Parameters: 
+ * stat: pointer to the stat to be freed
+ * 
+ * Returns:
+ *  SUCCESS on success, FAILURE if an error occurs.
+ */
+int free_stats(stats_t *stat);
+
+/*
+ * Frees a global stat
  *
  * Parameters: 
- * sh: pointer to the stats hash table to be freed
+ * stat: pointer to the global stat to be freed
  * 
  * Returns:
  *  SUCCESS on success, FAILURE if an error occurs.
  */
-int free_stats(stats_hash_t *sh);
+int free_stats_global(stats_global_t *stat);
```
Did we ever reach concesus on whether the free functions should free the whole table or just individual stats? I'm still unsure of a situation where you want to free an individual stat, and not the whole table, so we might as well do that?

sumedhsarwate replied:
I'll change it to freeing the entire table for now, but it might be good to add a function for freeing the individual stat later

carlosazpurua1004 commented on src/game-state/src/stats.c:
```diff
@@ -1,14 +1,16 @@
+#define MAX_NAME_LENGTH 50
 
 /* See stats.h */
-int stats_init(stats_t *s, char *stats_name, double init)
+int stats_init(stats_t *s, stats_global_t *stats_name, double init)
 {
     printf("stats_init: function not yet implemented\n");
     return 0; // still needs to be implemented
 }
 
 /* See stats.h */
-stats_t *stats_new(char *stats_name, double init)
+stats_t *stats_new(stats_global_t *stats_name, double init)
```
I made other changes to the headers of these functions while writing them, I think it would be best to just undo changes here to avoid merge conflicts since these changes since they don't affect the rest of the code.

carlosazpurua1004 commented on include/game-state/stats.h:
```diff
@@ -66,30 +83,59 @@ typedef struct effects{
 
 typedef struct effects effects_hash_t;
 
+
+/*
+ * Initializes a global stat with max value stated
+ *
+ * Parameters:
+ *  s: A global stat struct Must point to already allocated memory.
+ *  name: the unique string ID that the stat has
+ *  max: maximal value of the stat
+ *
+ * Returns:
+ *  SUCCESS on success, FAILURE if an error occurs.
+ */
+
+int stats_global_init(stats_global_t *s, char *name, double max);
+
+
 /*
  * Initializes a Stat with specified value and modifier 0
  *
  * Parameters:
  *  s: A stats struct Must point to already allocated memory.
- *  stats_name: the unique string ID of the stat
+ *  stat: the pointer to the global stat struct
  *  init: starting value
  *
  * Returns:
  *  SUCCESS on success, FAILURE if an error occurs.
  */
-int stats_init(stats_t *s, char *stats_name, double init);
+int stats_init(stats_t *s, stats_global_t *stat, double init);
```
See later comment about header.

carlosazpurua1004 commented on include/game-state/stats.h:
```diff
@@ -66,30 +83,59 @@ typedef struct effects{
  * Initializes a Stat with specified value and modifier 0
  *
  * Parameters:
  *  s: A stats struct Must point to already allocated memory.
- *  stats_name: the unique string ID of the stat
+ *  stat: the pointer to the global stat struct
  *  init: starting value
  *
  * Returns:
  *  SUCCESS on success, FAILURE if an error occurs.
  */
-int stats_init(stats_t *s, char *stats_name, double init);
+int stats_init(stats_t *s, stats_global_t *stat, double init);
+
+/*
+ * Allocates a new global stat
+ *
+ * Parameters:
+ * name: the unique string ID to be given to the stat
+ * max: maximal value this stat could have
+ * 
+ * Returns:
+ *  Pointer to allocated global stats struct
+ */
+
+stats_global_t *stats_global_new(char *name, double max);
 
 /*
  * Allocates a new stat
  *
  * Parameters:
- * stats_name: the unique string ID to be given to the stat
+ * stat: the pointer to the global stat struct.
  * init: starting value
  * 
  * Returns:
  *  Pointer to allocated stats struct
  */
-stats_t *stats_new(char *stats_name, double init);
+stats_t *stats_new(stats_global_t *stat, double init);
```
See later comment about header.

carlosazpurua1004 commented on src/game-state/src/stats.c:
```diff
@@ -43,17 +43,42 @@ double get_stat_mod(stats_hash_t *sh, char *stat)
-int add_stat_player(stats_hash_t *sh, stats_t *s)
+int add_stat_player(stats_hash_t **sh, stats_t *s)
 {
-    printf("add_stat: function not yet implemented\n");
-    return 0; // still needs to be implemented
+    stats_t *check;
+    
+    HASH_FIND(hh, *sh, s->global, sizeof(stats_global_t), check);
```
nvm

sumedhsarwate commented:
I've updated the stats struct so that we now have a string key and I've updated the add_stat functions and the tests accordingly.

sumedhsarwate commented:
Summary: The global stats struct was added to dev along with the add_stat and display_stat functions and their tests. The stats struct has also been updated so that its key is a string instead of a pointer to a global stats struct.

ecoble commented:
Issue Score: ✔️++
Comments:
Great work, especially with the discussions on this PR!
<!-- {"score": "check-plus-plus"} -->
